lib/WUI/nhttp/req_parser.cpp

```diff
@@ -0,0 +1,177 @@
+            return ExecutionControl::Continue;
+        }
+    case Names::Boundary:
+        if (boundary_size < boundary_buff.size()) {
+            boundary_buff[boundary_size++] = event.payload;
+            return ExecutionControl::Continue;
+        } else {
+            error_code = Status::RequestHeaderFieldsTooLarge;
+            return ExecutionControl::Continue;
+        }
+    case Names::XApiKey:
+        // No API key -> can't authenticate.
+        if (!api_key) {
+            auth_status = false;
+        }
+        // The first character, just starting.
+        if (holds_alternative<monostate>(auth_status)) {
+            auth_status = static_cast<uint8_t>(0);
+        }
+        if (uint8_t *idx = get_if<uint8_t>(&auth_status); idx != nullptr) {
+            const size_t key_len = strlen(api_key);
+            if (*idx >= key_len || api_key[*idx] != event.payload) {
+                auth_status = false;
+            } else if (++*idx == key_len) {
+                auth_status = true;
+            }
+        } else {
+            auth_status = false;
+        }
+        return ExecutionControl::Continue;
+    case Names::ContentLength:
+        if (!content_length.has_value()) {
+            content_length = 0;
+        }
+        *content_length = 10 * *content_length + (event.payload - '0');
+        return ExecutionControl::Continue;
+    case Names::Version:
+        switch (event.payload) {
+        case '.':
+            version_major = version_minor;
+            version_minor = 0;
+            return ExecutionControl::Continue;
+        case '0' ... '9':
+            version_minor = 10 * version_minor + (event.payload - '0');
+            return ExecutionControl::Continue;
+        }
+    case Names::ConnectionClose:
+        connection = Connection::Close;
+        break;
+    case Names::ConnectionKeepAlive:
+        connection = Connection::KeepAlive;
+        break;
+    case Names::Body:
+        done = true;
+        // Yes, really don't stop. Eath the \n too!
+        return ExecutionControl::Continue;
+    }
+
+    return ExecutionControl::Continue;
+}
+
+Step RequestParser::step(string_view input, bool terminated_by_client, uint8_t *, size_t) {
+    if (done) {
+        return Step { 0, 0, Continue() };
+    }
+
+    api_key = server->get_api_key();
+    const auto [result, consumed] = consume(input);
+    api_key = nullptr;
+
+    if (!done && result == ExecutionControl::NoTransition) {
+        // Malformed request
+        error_code = Status::BadRequest;
+        done = true;
+    }
+
+    if (!done && terminated_by_client) {
+        // Incomplete request
+        error_code = Status::BadRequest;
+        done = true;
+    }
+
+    if (done) {
+        /*
+         * Note: we assume the array of selectors is terminated by one that
+         * handles unhandled/unknown requests ‒ one that returns 404 or so.
+         */
+        for (const Selector *const *selector = server->selectors();; selector++) {
```

Comment:
I'd recommend adding some comment between the two `;;` to avoid confusion (that it is not a typo but you really meant it that way)